### PR TITLE
Added State to Experiment

### PR DESCRIPTION
**Pull request by MarcusZuber**
Adds state to Experiment. Now running an already running experiemnt throws an exception.

### CODE REVIEW

**Comment by tfarago** [comment]
Also, please test this.

**Comment by MarcusZuber** [comment]
When using the `_state_value` it is quite ugly to check if an `abort()` finished (since the abort can not kill the future, but needs to be handled in the implementation of the frame gernarots). 
Therefore, I would prefer to check the state from the future as I did it in the first place to make sure, that really nothing is running in the background.

File: concert/tests/integration/test_experiments.py [hunk 1]
```diff
@@ -242,3 +264,35 @@ def test_attach_num_times(self):
...
+        self.assertEqual(exp.state, "running")
+        exp.abort().join()
+        self.assertEqual(exp.state, "aborted")
+        exp_handle.join()
+
+    def test_experiment_exception(self):
+        exp = ExperimentException(self.walker)
+        self.assertEqual(exp.state, "standby")
+        exp_handle = exp.run()
+        sleep(0.1)
```
Comment (tfarago) [inline concert/tests/integration/test_experiments.py:297]: Why do you need to sleep and not simply wait for `exp_handle.join()`?

**Comment by tfarago** [comment]
> When using the `_state_value` it is quite ugly to check if an `abort()` finished (since the abort can not kill the future, but needs to be handled in the implementation of the frame gernarots).
> Therefore, I would prefer to check the state from the future as I did it in the first place to make sure, that really nothing is running in the background.

Why? in the end of `abort` you can set the state to abort. Until you reach that point in the code the state is anyway undefined because half the acquisitions might be getting aborted and the other half not yet. It's actually exactly what you are doing in the commit. Also, why do you sleep there? You like those sleeps don't you? :smile:

File: concert/tests/integration/test_experiments.py (hunk 1 above)
Comment (MarcusZuber) [reply concert/tests/integration/test_experiments.py:297]: join will block until the state is back to standby and the `_state_value` is not set fast enough. Without it you have a race condition if the state was set fast enough (that also why I don't like the `_state_value`)

**Comment by MarcusZuber** [comment]
I hate those sleeps...
The commit with the abort works, if the user sets the state back from running. If this is not done correctly in the Acquisition the abort will block forever. And if I don't wait for the state change, it could be, that a 2nd run is started while the 1st one is still not completely aborted.
I still prefer the version with the checking of the future.

**Comment by tfarago** [comment]
Let me play with this a little on Monday.

**Comment by MarcusZuber** [comment]
What I could do, would be that I store the future in the class and check the state of the furture in the abort function and keep the state in "aborting" or something similar until it joins. But then, the difference of directly checking the future  state is quite little.

**Comment by tfarago** [comment]
FYI:I had to force-push because master was ahead.

**Comment by tfarago** [comment]
How about 60f479d5bd079e2103f3f78d150a633b2cd4a5fe? There is no way you can start before aborting is finished and I don't see how abort can block forever.

**Comment by MarcusZuber** [comment]
`Acquisition.abort().join()` will stop the generation of new frames, but what happens with the `Acquisition.producer`? I guess, it will still run and needs to be somehow notified about the abort. 
Also, in https://github.com/ufo-kit/concert/commit/60f479d5bd079e2103f3f78d150a633b2cd4a5fe the state will not change from aborting to aborted (which maybe can be done after the `acq.abort()`)

**Comment by tfarago** [comment]
> `Acquisition.abort().join()` will stop the generation of new frames, but what happens with the `Acquisition.producer`?

This is actually unrelated to this PR.

> the state will not change from aborting to aborted (which maybe can be done after the acq.abort())

Why should it? It's changed to `standby` automatically after `run` finishes. Do you really need to know about aborted state? Should the devices have `aborted` state? I would like to keep things consistent.

**Comment by tfarago** [comment]
> `Acquisition.abort().join()` will stop the generation of new frames, but what happens with the `Acquisition.producer`?

AFAIK the producer coroutine finishes as well. If you put it's code in try-finally the finally clause gets executed.

**Comment by MarcusZuber** [comment]
> Why should it? It's changed to `standby` automatically after `run` finishes. Do you really need to know about aborted state? Should the devices have `aborted` state? I would like to keep things consistent.

That's right.



> > `Acquisition.abort().join()` will stop the generation of new frames, but what happens with the `Acquisition.producer`?
> 
> This is actually unrelated to this PR.

Yes for this PR it is not so important. However, I think the abort should stop everything inside of the experiment and should somehow allow to check if it really stopped and not only stopping the generation of frames.



> > `Acquisition.abort().join()` will stop the generation of new frames, but what happens with the `Acquisition.producer`?
> 
> AFAIK the producer coroutine finishes as well. If you put it's code in try-finally the finally clause gets executed.

What do you think of splitting this in two functions? We could put this try-finally block in the Acquisition and the "user" provides a producerfunction (which will be in the try-part and can be the producer as it is used so far) and can provide an additional "abort-handler"-function which will be called in the finally-block?

**Comment by tfarago** [comment]
> Yes for this PR it is not so important. However, I think the abort should stop everything inside of the experiment and should somehow allow to check if it really stopped and not only stopping the generation of frames.

Then let's finish what was the purpose of this PR here and start a clean new issue/PR elsewhere.

> What do you think of splitting this in two functions? We could put this try-finally block in the Acquisition and the "user" provides a producerfunction (which will be in the try-part and can be the producer as it is used so far) and can provide an additional "abort-handler"-function which will be called in the finally-block?

Isn't this what `finish` is for?

**Comment by tfarago** [comment]
But yeah, that doesn't help on the finer acquisition level...

**Comment by MarcusZuber** [comment]
Ok. Can you merge it with your last change? I will create a new PR with this abort-handler after you merged. The we can discuss this there.

**Pull request merged**